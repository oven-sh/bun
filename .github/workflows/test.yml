name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
