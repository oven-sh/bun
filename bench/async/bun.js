import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
