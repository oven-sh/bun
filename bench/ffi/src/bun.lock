{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
