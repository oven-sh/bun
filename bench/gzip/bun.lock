{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
