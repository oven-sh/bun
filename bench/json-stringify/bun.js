import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
