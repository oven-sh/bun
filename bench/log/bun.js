import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
