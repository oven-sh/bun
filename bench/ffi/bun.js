import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
