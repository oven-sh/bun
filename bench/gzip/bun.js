import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
