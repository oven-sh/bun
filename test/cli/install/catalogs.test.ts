import { file, spawn, write } from "bun";
import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { exists } from "fs/promises";
import { VerdaccioRegistry, bunEnv, bunExe, runBunInstall, stderrForInstall } from "harness";
import { join } from "path";

var registry = new VerdaccioRegistry();

beforeAll(async () => {
  await registry.start();
});

afterAll(() => {
  registry.stop();
});

describe("basic", () => {
  async function createBasicCatalogMonorepo(packageDir: string, name: string) {
    const packageJson = {
      name,
      workspaces: {
        packages: ["packages/*"],
        catalog: {
          "no-deps": "2.0.0",
        },
        catalogs: {
          a: {
            "a-dep": "1.0.1",
          },
        },
      },
    };

    await Promise.all([
      write(join(packageDir, "package.json"), JSON.stringify(packageJson)),
      write(
        join(packageDir, "packages", "pkg1", "package.json"),
        JSON.stringify({
          name: "pkg1",
          dependencies: {
            "no-deps": "catalog:",
            "a-dep": "catalog:a",
          },
        }),
      ),
    ]);

    return packageJson;
  }
  test("both catalog and catalogs", async () => {
    const { packageDir } = await registry.createTestDir();

    await createBasicCatalogMonorepo(packageDir, "catalog-basic-1");

    await runBunInstall(bunEnv, packageDir);

    expect(await file(join(packageDir, "node_modules", "no-deps", "package.json")).json()).toEqual({
      name: "no-deps",
      version: "2.0.0",
    });

    expect(await file(join(packageDir, "node_modules", "a-dep", "package.json")).json()).toEqual({
      name: "a-dep",
      version: "1.0.1",
    });

    // another install does not save the lockfile
    await runBunInstall(bunEnv, packageDir, { savesLockfile: false });
  });

  for (const binaryLockfile of [true, false]) {
    test(`detect changes (${binaryLockfile ? "bun.lockb" : "bun.lock"})`, async () => {
      const { packageDir } = await registry.createTestDir({ saveTextLockfile: !binaryLockfile });
      const packageJson = await createBasicCatalogMonorepo(packageDir, "catalog-basic-2");
      let { err } = await runBunInstall(bunEnv, packageDir);
      expect(err).toContain("Saved lockfile");

      const initialLockfile = !binaryLockfile
        ? (await file(join(packageDir, "bun.lock")).text()).replaceAll(/localhost:\d+/g, "localhost:1234")
        : undefined;

      if (!binaryLockfile) {
        expect(initialLockfile).toMatchSnapshot();
      } else {
        expect(await exists(join(packageDir, "bun.lockb"))).toBeTrue();
      }

      expect(await file(join(packageDir, "node_modules", "no-deps", "package.json")).json()).toEqual({
        name: "no-deps",
        version: "2.0.0",
      });
      expect(await file(join(packageDir, "node_modules", "a-dep", "package.json")).json()).toEqual({
        name: "a-dep",
        version: "1.0.1",
      });

      // update catalog
      packageJson.workspaces.catalog["no-deps"] = "1.0.0";
      await write(join(packageDir, "package.json"), JSON.stringify(packageJson));
      ({ err } = await runBunInstall(bunEnv, packageDir, { savesLockfile: true }));
      expect(err).toContain("Saved lockfile");

      if (!binaryLockfile) {
        const newLockfile = (await file(join(packageDir, "bun.lock")).text()).replaceAll(
          /localhost:\d+/g,
          "localhost:1234",
        );

        expect(newLockfile).not.toEqual(initialLockfile);
        expect(newLockfile).toMatchSnapshot();
      } else {
        expect(await exists(join(packageDir, "bun.lockb"))).toBeTrue();
      }

      expect(await file(join(packageDir, "node_modules", "no-deps", "package.json")).json()).toEqual({
        name: "no-deps",
        version: "1.0.0",
      });
      expect(await file(join(packageDir, "node_modules", "a-dep", "package.json")).json()).toEqual({
        name: "a-dep",
        version: "1.0.1",
      });

      // update catalogs
      packageJson.workspaces.catalogs.a["a-dep"] = "1.0.10";
      await write(join(packageDir, "package.json"), JSON.stringify(packageJson));
      ({ err } = await runBunInstall(bunEnv, packageDir, { savesLockfile: true }));
      expect(err).toContain("Saved lockfile");

      if (!binaryLockfile) {
        const newLockfile = (await file(join(packageDir, "bun.lock")).text()).replaceAll(
          /localhost:\d+/g,
          "localhost:1234",
        );

        expect(newLockfile).not.toEqual(initialLockfile);
        expect(newLockfile).toMatchSnapshot();
      } else {
        expect(await exists(join(packageDir, "bun.lockb"))).toBeTrue();
      }

      expect(await file(join(packageDir, "node_modules", "no-deps", "package.json")).json()).toEqual({
        name: "no-deps",
        version: "1.0.0",
      });
      expect(await file(join(packageDir, "node_modules", "a-dep", "package.json")).json()).toEqual({
        name: "a-dep",
        version: "1.0.10",
      });
    });
  }
});

describe("errors", () => {
  test("fails gracefully when no catalog is found for a package", async () => {
    const { packageDir, packageJson } = await registry.createTestDir();

    await write(
      packageJson,
      JSON.stringify({
        name: "catalog-error-1",
        workspaces: {
          // empty, any catalog should fail to resolve
          catalog: {},
          catalogs: {},
        },
        dependencies: {
          "no-deps": "catalog:",

          // longer than 8
          "a-dep": "catalog:aaaaaaaaaaaaaaaaa",
        },
      }),
    );

    const { stdout, stderr, exited } = spawn({
      cmd: [bunExe(), "install"],
      cwd: packageDir,
      stdout: "pipe",
      stderr: "pipe",
      env: bunEnv,
    });

    const out = await Bun.readableStreamToText(stdout);
    const err = stderrForInstall(await Bun.readableStreamToText(stderr));

    expect(err).toContain("no-deps@catalog: failed to resolve");
    expect(err).toContain("a-dep@catalog:aaaaaaaaaaaaaaaaa failed to resolve");
  });

  test("invalid dependency version", async () => {
    const { packageDir, packageJson } = await registry.createTestDir();
    await write(
      packageJson,
      JSON.stringify({
        name: "catalog-error-2",
        workspaces: {
          catalog: {
            "no-deps": ".:",
          },
        },
        dependencies: {
          "no-deps": "catalog:",
        },
      }),
    );

    const { stdout, stderr, exited } = spawn({
      cmd: [bunExe(), "install"],
      cwd: packageDir,
      stdout: "pipe",
      stderr: "pipe",
      env: bunEnv,
    });

    const out = await Bun.readableStreamToText(stdout);
    const err = stderrForInstall(await Bun.readableStreamToText(stderr));

    expect(err).toContain("no-deps@catalog: failed to resolve");
  });
});
