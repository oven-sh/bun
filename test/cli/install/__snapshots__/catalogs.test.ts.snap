// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`basic detect changes (bun.lock) 1`] = `
"{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "catalog-basic-2",
    },
    "packages/pkg1": {
      "name": "pkg1",
      "dependencies": {
        "a-dep": "catalog:a",
        "no-deps": "catalog:",
      },
    },
  },
  "catalog": {
    "no-deps": "2.0.0",
  },
  "catalogs": {
    "a": {
      "a-dep": "1.0.1",
    },
  },
  "packages": {
    "a-dep": ["a-dep@1.0.1", "http://localhost:1234/a-dep/-/a-dep-1.0.1.tgz", {}, "sha512-6nmTaPgO2U/uOODqOhbjbnaB4xHuZ+UB7AjKUA3g2dT4WRWeNxgp0dC8Db4swXSnO5/uLLUdFmUJKINNBO/3wg=="],

    "no-deps": ["no-deps@2.0.0", "http://localhost:1234/no-deps/-/no-deps-2.0.0.tgz", {}, "sha512-W3duJKZPcMIG5rA1io5cSK/bhW9rWFz+jFxZsKS/3suK4qHDkQNxUTEXee9/hTaAoDCeHWQqogukWYKzfr6X4g=="],

    "pkg1": ["pkg1@workspace:packages/pkg1"],
  }
}
"
`;

exports[`basic detect changes (bun.lock) 2`] = `
"{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "catalog-basic-2",
    },
    "packages/pkg1": {
      "name": "pkg1",
      "dependencies": {
        "a-dep": "catalog:a",
        "no-deps": "catalog:",
      },
    },
  },
  "catalog": {
    "no-deps": "1.0.0",
  },
  "catalogs": {
    "a": {
      "a-dep": "1.0.1",
    },
  },
  "packages": {
    "a-dep": ["a-dep@1.0.1", "http://localhost:1234/a-dep/-/a-dep-1.0.1.tgz", {}, "sha512-6nmTaPgO2U/uOODqOhbjbnaB4xHuZ+UB7AjKUA3g2dT4WRWeNxgp0dC8Db4swXSnO5/uLLUdFmUJKINNBO/3wg=="],

    "no-deps": ["no-deps@1.0.0", "http://localhost:1234/no-deps/-/no-deps-1.0.0.tgz", {}, "sha512-v4w12JRjUGvfHDUP8vFDwu0gUWu04j0cv9hLb1Abf9VdaXu4XcrddYFTMVBVvmldKViGWH7jrb6xPJRF0wq6gw=="],

    "pkg1": ["pkg1@workspace:packages/pkg1"],
  }
}
"
`;

exports[`basic detect changes (bun.lock) 3`] = `
"{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "catalog-basic-2",
    },
    "packages/pkg1": {
      "name": "pkg1",
      "dependencies": {
        "a-dep": "catalog:a",
        "no-deps": "catalog:",
      },
    },
  },
  "catalog": {
    "no-deps": "1.0.0",
  },
  "catalogs": {
    "a": {
      "a-dep": "1.0.10",
    },
  },
  "packages": {
    "a-dep": ["a-dep@1.0.10", "http://localhost:1234/a-dep/-/a-dep-1.0.10.tgz", {}, "sha512-NeQ6Ql9jRW8V+VOiVb+PSQAYOvVoSimW+tXaR0CoJk4kM9RIk/XlAUGCsNtn5XqjlDO4hcH8NcyaL507InevEg=="],

    "no-deps": ["no-deps@1.0.0", "http://localhost:1234/no-deps/-/no-deps-1.0.0.tgz", {}, "sha512-v4w12JRjUGvfHDUP8vFDwu0gUWu04j0cv9hLb1Abf9VdaXu4XcrddYFTMVBVvmldKViGWH7jrb6xPJRF0wq6gw=="],

    "pkg1": ["pkg1@workspace:packages/pkg1"],
  }
}
"
`;
